Firestore/src/WriteBatch.php

```diff
@@ -726,67 +687,58 @@ private function filterFields(array $fields)
                         ? $inArray
                         : !$this->isAssoc($value);
 
-                    $fields[$key] = $filterSentinels($value, $currentPath, $localInArray);
+                    $fields[$key] = $filterFn($value, $currentPath, $localInArray);
                     if (empty($fields[$key])) {
                         unset($fields[$key]);
                     }
                 } else {
-                    if (FieldValue::isSentinelValue($value)) {
-                        $sentinels[$value][] = $currentPath;
-                        if ($value === FieldValue::serverTimestamp() && $inArray) {
-                            $flags['timestampInArray'] = true;
+                    if ($value instanceof FieldValueInterface) {
+                        $value->setFieldPath($currentPath);
+                        $sentinels[] = $value;
+
+                        // Sentinels cannot be used within a non-associative array.
+                        if ($inArray) {
+                            throw new \InvalidArgumentException(sprintf(
+                                '%s values cannot be used anywhere within a non-associative array value.',
+                                FieldValue::class
+                            ));
+                        }
+
+                        // Delete cannot be nested in update-paths
+                        // (i.e. the only case where `$inputPaths` would be available)
+                        $illegalNestedDelete = $inputPaths
+                            && $value instanceof DeleteFieldValue
+                            && !in_array($currentPath, $inputPaths);
+
+                        if ($illegalNestedDelete) {
+                            throw new \InvalidArgumentException(sprintf(
+                                '%s::deleteField() values cannot be nested.',
+                                FieldValue::class
+                            ));
                         }
 
+                        // Remove the sentinel value from the fields array.
                         unset($fields[$key]);
+
+                        // Record the sentinel type.
+                        if (!in_array(get_class($value), $metadata['types'])) {
+                            $metadata['types'][] = get_class($value);
+                        }
+
+                        // Record whether the mutation contains a transform value.
+                        if (!$metadata['hasTransform'] && $value instanceof DocumentTransformInterface) {
+                            $metadata['hasTransform'] = true;
+                        }
```

Comment:
Note that you also need to validate that ArrayUnion/ArrayRemove filters do not contain nested transforms (there are essentially in_array).